prow/github/client.go

```diff
@@ -63,6 +64,99 @@ const (
 	initialDelay  = 2 * time.Second
 )
 
+// Interface for how prow interacts with the http client, which we may throttle.
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Inteface for how prow interacts with the graphql client, which we may throttle.
+type gqlClient interface {
+	Query(ctx context.Context, q interface{}, vars map[string]interface{}) error
+}
+
+// throttler sets a ceiling on the rate of github requests.
+// Configure with Client.Throttle()
+type throttler struct {
+	ticker   *time.Ticker
+	throttle chan time.Time
+	Http     httpClient
+	Graph    gqlClient
+	slow     int32
+}
+
+func (t *throttler) Wait() {
```

Comment:
This function isn't thread safe. Since this function doesn't hold the mutex another thread could call `Throttle(0,0)` while this function is executing and `t.throttle` could be read while it is being written.

Reply:
Done